### PR TITLE
set empty map for 'params'  in binding

### PR DESCRIPTION
**Pull request by noiseoverip**
without this. calling addParam() fails

### CODE REVIEW

File: src/main/groovy/com/lesfurets/jenkins/unit/BasePipelineTest.groovy [hunk 1]
```diff
@@ -291,7 +291,9 @@ abstract class BasePipelineTest {
         ])
         binding.setVariable('docker', new DockerMock())
         binding.setVariable('env', [:])
+        binding.setVariable('params', [:])
         binding.setVariable('scm', [:])
+
```
Comment (nre-ableton) [inline src/main/groovy/com/lesfurets/jenkins/unit/BasePipelineTest.groovy:296]:
```suggestion
```

**Review by nre-ableton — COMMENTED** [review]
I need to play around with this branch a bit, as I'm unfamiliar with the `params` singleton, but I at least added a suggestion to fix a whitespace error.

**Comment by nre-ableton** [comment]
I'm not convinced that this change is needed, though it is fairly innocuous. `BasePipelineTest.addParam()` can handle the case where `params` has not yet been added to the binding, see: https://github.com/jenkinsci/JenkinsPipelineUnit/blob/350a6e51fd58366c4aa302b6cbcc8782b50c5225/src/main/groovy/com/lesfurets/jenkins/unit/BasePipelineTest.groovy#L435-L444

Also, there is a test Jenkinsfile for this feature: `src/test/jenkins/jenkinsfiles/Params_Jenkinsfile`

Can you please show an example of a Jenkinsfile which fails without this patch?

**Comment by stchar** [comment]
@noiseoverip, could you please add a test to reproduce the issue (for scritped and declarative pipelines)

**Comment by noiseoverip** [comment]
Hi,
`BasePipelineTest.addParam()` doesn't help because it will throw exception when `params` is not initialized due to behavior of `getVariable()`. 

```    
public Object getVariable(String name) {
        if (variables == null)
            throw new MissingPropertyException(name, this.getClass());

        Object result = variables.get(name);

        if (result == null && !variables.containsKey(name)) {
            throw new MissingPropertyException(name, this.getClass());
        }

        return result;
    }
```

I think it is reasonable to initialize it by default since the same is done to `env` var and that would more closely match Jenkins behavior since script can call `params.FOO` under any circumstance (parameterized, non-parameterized) job and that doesn't throw any exceptions just returns NULL when param doesn't have any value. And yes i can create unit tests, just need to find time for this

**Comment by nre-ableton** [comment]
Perhaps I am missing something obvious here, but I added a test based on your suggestion, in `TestRegression.groovy`:

```groovy
    @Test
    void testParamsMapInitialized() throws Exception {
        def script = runScript("job/params.jenkins")
        script.execute()
        super.testNonRegression("params")
    }
```

...with the corresponding Jenkinsfile:

```groovy
def execute() {
    echo params.getClass().toString()
    echo(params.FOO)
}
return this
```

The test fails, even if this patch is applied (condensed stacktrace):

```
class java.lang.Class

groovy.lang.MissingPropertyException: No such property: FOO for class: params

	at groovy.lang.MetaClassImpl.invokeStaticMissingProperty(MetaClassImpl.java:1001)
        ...
	at ___cps.transform___(Native Method)
	at com.lesfurets.jenkins.unit.cps.PipelineTestHelperCPS.callMethod(PipelineTestHelperCPS.groovy:32)
        ...
	at com.lesfurets.jenkins.TestRegression.testParamsMapInitialized(TestRegression.groovy:33)
        ...
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:53)
```

Note that the echo statement thinks that `params` is a `Class` object, not a `Map`.

**Comment by noiseoverip** [comment]
I believe it fails because you don't have:
```    
@Override
@Before
    void setUp() throws Exception {
        super.setUp()
}
```

Since all the helper stuff doesn't get initialized if you don't do this

**Comment by noiseoverip** [comment]
the stack trace you printed is from a run without a patch. Adding that patch fixes it though of course the test it self fails due to callstack now being different. I have pushed a simple unit test to prove my point. That unit test fails without the patch

**Comment by JorisBolsens** [comment]
> Perhaps I am missing something obvious here, but I added a test based on your suggestion, in `TestRegression.groovy`:
> 
> ```groovy
>     @Test
>     void testParamsMapInitialized() throws Exception {
>         def script = runScript("job/params.jenkins")
>         script.execute()
>         super.testNonRegression("params")
>     }
> ```
> 
> ...with the corresponding Jenkinsfile:
> 
> ```groovy
> def execute() {
>     echo params.getClass().toString()
>     echo(params.FOO)
> }
> return this
> ```
> 
> The test fails, even if this patch is applied (condensed stacktrace):
> 
> ```
> class java.lang.Class
> 
> groovy.lang.MissingPropertyException: No such property: FOO for class: params
> 
> 	at groovy.lang.MetaClassImpl.invokeStaticMissingProperty(MetaClassImpl.java:1001)
>         ...
> 	at ___cps.transform___(Native Method)
> 	at com.lesfurets.jenkins.unit.cps.PipelineTestHelperCPS.callMethod(PipelineTestHelperCPS.groovy:32)
>         ...
> 	at com.lesfurets.jenkins.TestRegression.testParamsMapInitialized(TestRegression.groovy:33)
>         ...
> 	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:53)
> ```
> 
> Note that the echo statement thinks that `params` is a `Class` object, not a `Map`.

The issue is that calling `addParam('FOO','x')` would throw an exception unless you first initialize `params` to `[:]` as you do with `env`

Your test fails with/without the patch, as it is testing something different. If you were to try and have your test pass like so:
```groovy
     @Test
     void testParamsMapInitialized() throws Exception {
         def script = runScript("job/params.jenkins")
         addParam('FOO','x')
         script.execute()
         super.testNonRegression("params")
     }
 ```
 without this patch, this test would fail as you will get exception when calling `addParam`

**Comment by stchar** [comment]
I think this fix is outdated as `master` already has a case to handle empty `params`
https://github.com/jenkinsci/JenkinsPipelineUnit/blob/master/src/main/groovy/com/lesfurets/jenkins/unit/BasePipelineTest.groovy#L435-L438

**Comment by stchar** [comment]
Will be released in v1.9

**Pull request closed without merging**